File: python/llm_sketchkit/summary.py
# SPDX-License-Identifier: Apache-2.0
# Code authors: Vijay Erramilli and Codex
"""Window-scoped exchange of existing sketches; see spec/summary.md."""

from __future__ import annotations

import base64
import binascii
import json
import re
from dataclasses import asdict, dataclass
from typing import Any

from google.protobuf.message import DecodeError  # type: ignore[import-untyped]

from . import _proto, bloom, frequentitems, hllpp, minhash

MAX_BYTES = 8 << 20
_MAX_INT = (1 << 63) - 1
_IDENTIFIER = re.compile(r"[A-Za-z0-9._:-]{1,128}", re.ASCII)


class SummaryError(ValueError):
    """Malformed, incompatible, conflicting, or oversized summary input."""


@dataclass
class Payload:
    data: bytes
    kind: str


@dataclass
class Envelope:
    accounting_id: str
    counters: dict[str, int]
    emitted_at_unix_nano: int
    epoch: str
    key_id: str
    observed_end_unix_nano: int
    observed_start_unix_nano: int
    producer_id: str
    scope_id: str
    sequence: int
    sketches: dict[str, Payload]
    version: int
    window_duration_unix_nano: int
    window_start_unix_nano: int

    def validate(self) -> None:
        for value in (
            self.version, self.sequence, self.emitted_at_unix_nano,
            self.observed_start_unix_nano, self.observed_end_unix_nano,
            self.window_duration_unix_nano, self.window_start_unix_nano,
        ):
            if type(value) is not int or not 0 <= value <= _MAX_INT:
                raise SummaryError("invalid summary integer")
        if self.version != 1 or self.sequence == 0:
            raise SummaryError("invalid summary version or sequence")
        for ident in (
            self.accounting_id, self.epoch, self.key_id,
            self.producer_id, self.scope_id,
        ):
            _check_identifier(ident)
        start = self.window_start_unix_nano
        duration = self.window_duration_unix_nano
        if (
            not 0 < duration <= 86_400_000_000_000
            or start + duration > _MAX_INT
            or start % duration != 0
            or not start <= self.observed_start_unix_nano
            <= self.observed_end_unix_nano <= start + duration
            or self.emitted_at_unix_nano < self.observed_end_unix_nano
        ):
            raise SummaryError("invalid summary observation interval")
        if (
            type(self.counters) is not dict or len(self.counters) > 128
            or type(self.sketches) is not dict or len(self.sketches) > 16
        ):
            raise SummaryError("invalid summary payload count")
        for name, count in self.counters.items():
            _check_identifier(name)
            if type(count) is not int or not 0 <= count <= _MAX_INT:
                raise SummaryError("invalid summary counter")
        size = 0
        for name, payload in self.sketches.items():
            _check_identifier(name)
            if not isinstance(payload, Payload) or type(payload.data) is not bytes:
                raise SummaryError("invalid summary sketch")
            size += len(payload.data)
            if size > MAX_BYTES or _merge_payload(payload).data != payload.data:
                raise SummaryError("invalid summary sketch state")

    def marshal_binary(self) -> bytes:
        """Return canonical JSON without mutating the receiver, including on error."""
        self.validate()
        document = asdict(self)
        document["sketches"] = {
            name: {"data": base64.b64encode(p.data).decode("ascii"), "kind": p.kind}
            for name, p in self.sketches.items()
        }
        data = json.dumps(document, sort_keys=True, separators=(",", ":")).encode()
        if len(data) > MAX_BYTES:
            raise SummaryError("summary exceeds size limit")
        return data

    @classmethod
    def parse(cls, data: bytes) -> Envelope:
        """Reject unknown fields, duplicates, noncanonical JSON, and invalid state."""
        if len(data) > MAX_BYTES:
            raise SummaryError("summary exceeds size limit")
        try:
            document: Any = json.loads(data)
            payloads = document["sketches"]
            if type(payloads) is not dict:
                raise SummaryError("invalid summary sketches")
            document["sketches"] = {
                name: Payload(
                    data=base64.b64decode(p["data"], validate=True), kind=p["kind"]
                )
                for name, p in payloads.items()
                if type(p) is dict and set(p) == {"data", "kind"}
            }
            if len(document["sketches"]) != len(payloads):
                raise SummaryError("invalid summary payload")
            envelope = cls(**document)
            if envelope.marshal_binary() != data:
                raise SummaryError("noncanonical summary JSON")
            return envelope
        except (ValueError, TypeError, KeyError, AttributeError, RecursionError) as exc:
            raise SummaryError("invalid summary JSON or state") from exc


def compatible(a: Envelope, b: Envelope) -> None:
    """Check comparison compatibility; window starts may differ. No mutation."""
    a.validate()
    b.validate()
    if (
        a.scope_id != b.scope_id or a.accounting_id != b.accounting_id
        or a.key_id != b.key_id
        or a.window_duration_unix_nano != b.window_duration_unix_nano
        or a.counters.keys() != b.counters.keys()
        or a.sketches.keys() != b.sketches.keys()
    ):
        raise SummaryError("incompatible summary measurement contract")
    for name, payload in a.sketches.items():
        other = b.sketches[name]
        left = _proto.parse_sketch(payload.data).metadata
        right = _proto.parse_sketch(other.data).metadata
        left.ClearField("representation_mode")
        right.ClearField("representation_mode")
        if payload.kind != other.kind or left != right:
            raise SummaryError("incompatible summary sketch metadata")


@dataclass
class Source:
    producer_id: str
    epoch: str
    sequence: int
    observed_start_unix_nano: int
    observed_end_unix_nano: int


@dataclass
class Result:
    counters: dict[str, int]
    sketches: dict[str, Payload]
    sources: list[Source]
    missing: list[str]
    partial: list[str]


def combine(inputs: list[Envelope], expected: list[str]) -> Result:
    """Rebuild one window from snapshots of declared disjoint input owners.

    All state returned is new. Errors do not mutate inputs or expose a partial
    result. Missing/partial producers are reported, not treated as measured zero.
    """
    if len(inputs) > 1024 or not 0 < len(expected) <= 128:
        raise SummaryError("invalid summary batch size")
    for name in expected:
        _check_identifier(name)
    if len(set(expected)) != len(expected):
        raise SummaryError("duplicate expected producer")
    encoded: list[tuple[Envelope, bytes]] = []
    size = 0
    for doc in inputs:
        data = doc.marshal_binary()
        size += len(data)
        if size > 64 << 20:
            raise SummaryError("summary batch exceeds size limit")
        if doc.producer_id not in expected:
            raise SummaryError("unexpected summary producer")
        encoded.append((doc, data))
    encoded.sort(key=lambda pair: (
        pair[0].producer_id, pair[0].epoch, pair[0].sequence
    ))
    selected: list[Envelope] = []
    previous_bytes = b""
    for doc, data in encoded:
        if selected:
            if doc.window_start_unix_nano != selected[0].window_start_unix_nano:
                raise SummaryError("cannot combine different windows")
            compatible(selected[0], doc)
            last = selected[-1]
            if (last.producer_id, last.epoch) == (doc.producer_id, doc.epoch):
                if last.sequence == doc.sequence:
                    if previous_bytes != data:
                        raise SummaryError("conflicting summary sequence")
                    continue
                if (
                    doc.observed_start_unix_nano != last.observed_start_unix_nano
                    or doc.observed_end_unix_nano < last.observed_end_unix_nano
                    or any(doc.counters[k] < v for k, v in last.counters.items())
                ):
                    raise SummaryError("summary observation or counter regressed")
                selected[-1] = doc
                previous_bytes = data
                continue
        selected.append(doc)
        previous_bytes = data
    result = Result({}, {}, [], [], [])
    intervals: dict[str, list[Envelope]] = {}
    for doc in selected:
        intervals.setdefault(doc.producer_id, []).append(doc)
        result.sources.append(Source(
            doc.producer_id, doc.epoch, doc.sequence,
            doc.observed_start_unix_nano, doc.observed_end_unix_nano,
        ))
        for name, value in doc.counters.items():
            total = result.counters.get(name, 0) + value
            if total > _MAX_INT:
                raise SummaryError("combined counter overflow")
            result.counters[name] = total
        for name, payload in doc.sketches.items():
            previous = result.sketches.get(name)
            result.sketches[name] = (
                _merge_payload(payload) if previous is None
                else _merge_payload(previous, payload)
            )
    for name in sorted(expected):
        if name not in intervals:
            result.missing.append(name)
            continue
        parts = sorted(intervals[name], key=lambda doc: doc.observed_start_unix_nano)
        end = parts[0].window_start_unix_nano
        partial = False
        for part in parts:
            if part.observed_start_unix_nano < end:
                raise SummaryError("overlapping producer epochs")
            partial = partial or part.observed_start_unix_nano != end
            end = part.observed_end_unix_nano
        if partial or end != (
            parts[0].window_start_unix_nano + parts[0].window_duration_unix_nano
        ):
            result.partial.append(name)
    return result


def _check_identifier(value: str) -> None:
    if not isinstance(value, str) or _IDENTIFIER.fullmatch(value) is None:
        raise SummaryError("invalid summary identifier")


def _merge_payload(a: Payload, b: Payload | None = None) -> Payload:
    if b is not None and a.kind != b.kind:
        raise SummaryError("sketch kinds differ")
    try:
        if a.kind == "hllpp":
            h = hllpp.Sketch.parse(a.data)
            if b is not None:
                h.merge(hllpp.Sketch.parse(b.data))
            data = h.marshal_binary()
        elif a.kind == "frequent_items":
            f = frequentitems.Sketch.parse(a.data)
            if b is not None:
                f.merge(frequentitems.Sketch.parse(b.data))
            data = f.marshal_binary()
        elif a.kind == "bloom":
            bl = bloom.Sketch.parse(a.data)
            if b is not None:
                bl.merge(bloom.Sketch.parse(b.data))
            data = bl.marshal_binary()
        elif a.kind == "minhash":
            m = minhash.Sketch.parse(a.data)
            if b is not None:
                m.merge(minhash.Sketch.parse(b.data))
            data = m.marshal_binary()
        else:
            raise SummaryError("unknown summary sketch kind")
        return Payload(data, a.kind)
    except (ValueError, binascii.Error, DecodeError) as exc:
        raise SummaryError("invalid or incompatible summary sketch state") from exc

File: go/sketchkit/summary/summary.go
// SPDX-License-Identifier: Apache-2.0
// Code authors: Vijay Erramilli and Codex

// Package summary exchanges window-scoped counters and existing sketch state.
// Producer trust, disjoint input ownership, and secret distribution are external.
package summary

import (
	"bytes"
	"encoding/json"
	"errors"
	"math"
	"regexp"
	"slices"
	"sort"

	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/bloom"
	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/frequentitems"
	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/hllpp"
	sketchpb "github.com/llm-measurement/llm-sketchkit/go/sketchkit/internal/pb"
	"github.com/llm-measurement/llm-sketchkit/go/sketchkit/minhash"
	"google.golang.org/protobuf/proto"
)

const MaxBytes = 8 << 20
const maxBatchBytes = 64 << 20
const maxDuration = int64(24 * 60 * 60 * 1e9)

var identifier = regexp.MustCompile(`^[A-Za-z0-9._:-]{1,128}$`)

// Payload holds canonical protobuf bytes, base64-encoded by JSON.
type Payload struct {
	Data []byte `json:"data"`
	Kind string `json:"kind"`
}

// Envelope is a cumulative snapshot of one producer epoch within one window.
// Field order is lexical to match the canonical JSON contract in both languages.
type Envelope struct {
	AccountingID   string             `json:"accounting_id"`
	Counters       map[string]uint64  `json:"counters"`
	EmittedAt      int64              `json:"emitted_at_unix_nano"`
	Epoch          string             `json:"epoch"`
	KeyID          string             `json:"key_id"`
	ObservedEnd    int64              `json:"observed_end_unix_nano"`
	ObservedStart  int64              `json:"observed_start_unix_nano"`
	ProducerID     string             `json:"producer_id"`
	ScopeID        string             `json:"scope_id"`
	Sequence       uint64             `json:"sequence"`
	Sketches       map[string]Payload `json:"sketches"`
	Version        int                `json:"version"`
	WindowDuration int64              `json:"window_duration_unix_nano"`
	WindowStart    int64              `json:"window_start_unix_nano"`
}

func (e Envelope) Validate() error {
	if e.Version != 1 || e.Sequence == 0 || e.Sequence > math.MaxInt64 {
		return errors.New("invalid summary version or sequence")
	}
	for _, id := range []string{e.AccountingID, e.Epoch, e.KeyID, e.ProducerID, e.ScopeID} {
		if !identifier.MatchString(id) {
			return errors.New("invalid summary identifier")
		}
	}
	if e.WindowDuration <= 0 || e.WindowDuration > maxDuration || e.WindowStart < 0 ||
		e.WindowStart > math.MaxInt64-e.WindowDuration || e.WindowStart%e.WindowDuration != 0 ||
		e.ObservedStart < e.WindowStart || e.ObservedEnd < e.ObservedStart ||
		e.ObservedEnd > e.WindowStart+e.WindowDuration || e.EmittedAt < e.ObservedEnd {
		return errors.New("invalid summary observation interval")
	}
	if e.Counters == nil || len(e.Counters) > 128 || e.Sketches == nil || len(e.Sketches) > 16 {
		return errors.New("invalid summary payload count")
	}
	for name, count := range e.Counters {
		if !identifier.MatchString(name) || count > math.MaxInt64 {
			return errors.New("invalid summary counter")
		}
	}
	size := 0
	for name, payload := range e.Sketches {
		size += len(payload.Data)
		if !identifier.MatchString(name) || size > MaxBytes {
			return errors.New("invalid summary sketch")
		}
		canonical, err := mergePayload(payload, nil)
		if err != nil || !bytes.Equal(canonical.Data, payload.Data) {
			return errors.New("invalid summary sketch state")
		}
	}
	return nil
}

// MarshalBinary returns canonical JSON. Errors leave the receiver unchanged.
func (e Envelope) MarshalBinary() ([]byte, error) {
	if err := e.Validate(); err != nil {
		return nil, err
	}
	data, err := json.Marshal(e)
	if err != nil {
		return nil, err
	}
	if len(data) > MaxBytes {
		return nil, errors.New("summary exceeds size limit")
	}
	return data, nil
}

// Parse rejects unknown fields, duplicates, noncanonical JSON, and invalid state.
func Parse(data []byte) (Envelope, error) {
	var e Envelope
	if len(data) > MaxBytes {
		return e, errors.New("summary exceeds size limit")
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&e); err != nil {
		return Envelope{}, errors.New("invalid summary JSON")
	}
	canonical, err := e.MarshalBinary()
	if err != nil {
		return Envelope{}, err
	}
	if !bytes.Equal(canonical, data) {
		return Envelope{}, errors.New("noncanonical summary JSON")
	}
	return e, nil
}

// Compatible checks measurement and sketch compatibility for comparison.
// Window starts may differ; durations must match. It does not mutate either input.
func Compatible(a, b Envelope) error {
	if err := a.Validate(); err != nil {
		return err
	}
	if err := b.Validate(); err != nil {
		return err
	}
	if a.ScopeID != b.ScopeID || a.AccountingID != b.AccountingID || a.KeyID != b.KeyID ||
		a.WindowDuration != b.WindowDuration || !slices.Equal(names(a.Counters), names(b.Counters)) ||
		!slices.Equal(names(a.Sketches), names(b.Sketches)) {
		return errors.New("incompatible summary measurement contract")
	}
	for name, payload := range a.Sketches {
		other := b.Sketches[name]
		var left, right sketchpb.Sketch
		if err := proto.Unmarshal(payload.Data, &left); err != nil {
			return err
		}
		if err := proto.Unmarshal(other.Data, &right); err != nil {
			return err
		}
		left.Metadata.RepresentationMode = 0
		right.Metadata.RepresentationMode = 0
		if payload.Kind != other.Kind || !proto.Equal(left.Metadata, right.Metadata) {
			return errors.New("incompatible summary sketch metadata")
		}
	}
	return nil
}

type Source struct {
	ProducerID    string `json:"producer_id"`
	Epoch         string `json:"epoch"`
	Sequence      uint64 `json:"sequence"`
	ObservedStart int64  `json:"observed_start_unix_nano"`
	ObservedEnd   int64  `json:"observed_end_unix_nano"`
}

// Result is new combined state. Missing and Partial refer to expected producers.
type Result struct {
	Counters map[string]uint64  `json:"counters"`
	Sketches map[string]Payload `json:"sketches"`
	Sources  []Source           `json:"sources"`
	Missing  []string           `json:"missing"`
	Partial  []string           `json:"partial"`
}

// Combine selects the latest cumulative snapshot per producer/epoch and rebuilds
// one window. expected declares disjoint input owners, not merely allowed names.
// Errors return no partial result and never modify the supplied envelopes.
func Combine(input []Envelope, expected []string) (Result, error) {
	if len(input) > 1024 || len(expected) == 0 || len(expected) > 128 {
		return Result{}, errors.New("invalid summary batch size")
	}
	owners := make(map[string]bool, len(expected))
	for _, id := range expected {
		if !identifier.MatchString(id) {
			return Result{}, errors.New("invalid expected producer")
		}
		if _, ok := owners[id]; ok {
			return Result{}, errors.New("duplicate expected producer")
		}
		owners[id] = false
	}
	// Sorting a copy makes replacement and merge order independent of arrival order.
	docs := append([]Envelope(nil), input...)
	sort.Slice(docs, func(i, j int) bool {
		a, b := docs[i], docs[j]
		if a.ProducerID != b.ProducerID {
			return a.ProducerID < b.ProducerID
		}
		if a.Epoch != b.Epoch {
			return a.Epoch < b.Epoch
		}
		return a.Sequence < b.Sequence
	})
	selected := make([]Envelope, 0, len(docs))
	encodedBytes := 0
	var previousBytes []byte
	for _, doc := range docs {
		encoded, err := doc.MarshalBinary()
		if err != nil {
			return Result{}, err
		}
		encodedBytes += len(encoded)
		if encodedBytes > maxBatchBytes {
			return Result{}, errors.New("summary batch exceeds size limit")
		}
		if _, ok := owners[doc.ProducerID]; !ok {
			return Result{}, errors.New("unexpected summary producer")
		}
		if len(selected) > 0 {
			if doc.WindowStart != selected[0].WindowStart {
				return Result{}, errors.New("cannot combine different windows")
			}
			if err := Compatible(selected[0], doc); err != nil {
				return Result{}, err
			}
			last := selected[len(selected)-1]
			if last.ProducerID == doc.ProducerID && last.Epoch == doc.Epoch {
				if last.Sequence == doc.Sequence {
					if !bytes.Equal(previousBytes, encoded) {
						return Result{}, errors.New("conflicting summary sequence")
					}
					continue
				}
				if doc.ObservedStart != last.ObservedStart || doc.ObservedEnd < last.ObservedEnd {
					return Result{}, errors.New("summary observation regressed")
				}
				for name, value := range last.Counters {
					if doc.Counters[name] < value {
						return Result{}, errors.New("summary counter regressed")
					}
				}
				selected[len(selected)-1] = doc
				previousBytes = encoded
				continue
			}
		}
		selected = append(selected, doc)
		previousBytes = encoded
	}
	result := Result{Counters: map[string]uint64{}, Sketches: map[string]Payload{}, Sources: []Source{}, Missing: []string{}, Partial: []string{}}
	intervals := make(map[string][]Envelope)
	for _, doc := range selected {
		owners[doc.ProducerID] = true
		intervals[doc.ProducerID] = append(intervals[doc.ProducerID], doc)
		result.Sources = append(result.Sources, Source{doc.ProducerID, doc.Epoch, doc.Sequence, doc.ObservedStart, doc.ObservedEnd})
		for name, value := range doc.Counters {
			if result.Counters[name] > math.MaxInt64-value {
				return Result{}, errors.New("combined counter overflow")
			}
			result.Counters[name] += value
		}
		for name, payload := range doc.Sketches {
			var merged Payload
			var err error
			if previous, ok := result.Sketches[name]; ok {
				merged, err = mergePayload(previous, &payload)
			} else {
				merged, err = mergePayload(payload, nil)
			}
			if err != nil {
				return Result{}, err
			}
			result.Sketches[name] = merged
		}
	}
	for _, id := range names(owners) {
		if !owners[id] {
			result.Missing = append(result.Missing, id)
			continue
		}
		parts := intervals[id]
		sort.Slice(parts, func(i, j int) bool { return parts[i].ObservedStart < parts[j].ObservedStart })
		end := parts[0].WindowStart
		partial := false
		for _, part := range parts {
			if part.ObservedStart < end {
				return Result{}, errors.New("overlapping producer epochs")
			}
			partial = partial || part.ObservedStart != end
			end = part.ObservedEnd
		}
		if partial || end != parts[0].WindowStart+parts[0].WindowDuration {
			result.Partial = append(result.Partial, id)
		}
	}
	return result, nil
}

func names[V any](values map[string]V) []string {
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

func mergePayload(a Payload, b *Payload) (Payload, error) {
	if b != nil && a.Kind != b.Kind {
		return Payload{}, errors.New("sketch kinds differ")
	}
	var data []byte
	var err error
	switch a.Kind {
	case "hllpp":
		s, parseErr := hllpp.Parse(a.Data)
		if parseErr != nil {
			return Payload{}, parseErr
		}
		if b != nil {
			other, parseErr := hllpp.Parse(b.Data)
			if parseErr != nil {
				return Payload{}, parseErr
			}
			if err = s.Merge(other); err != nil {
				return Payload{}, err
			}
		}
		data, err = s.MarshalBinary()
	case "frequent_items":
		s, parseErr := frequentitems.Parse(a.Data)
		if parseErr != nil {
			return Payload{}, parseErr
		}
		if b != nil {
			other, parseErr := frequentitems.Parse(b.Data)
			if parseErr != nil {
				return Payload{}, parseErr
			}
			if err = s.Merge(other); err != nil {
				return Payload{}, err
			}
		}
		data, err = s.MarshalBinary()
	case "bloom":
		s, parseErr := bloom.Parse(a.Data)
		if parseErr != nil {
			return Payload{}, parseErr
		}
		if b != nil {
			other, parseErr := bloom.Parse(b.Data)
			if parseErr != nil {
				return Payload{}, parseErr
			}
			if err = s.Merge(other); err != nil {
				return Payload{}, err
			}
		}
		data, err = s.MarshalBinary()
	case "minhash":
		s, parseErr := minhash.Parse(a.Data)
		if parseErr != nil {
			return Payload{}, parseErr
		}
		if b != nil {
			other, parseErr := minhash.Parse(b.Data)
			if parseErr != nil {
				return Payload{}, parseErr
			}
			if err = s.Merge(other); err != nil {
				return Payload{}, err
			}
		}
		data, err = s.MarshalBinary()
	default:
		return Payload{}, errors.New("unknown summary sketch kind")
	}
	return Payload{Data: data, Kind: a.Kind}, err
}
